avocado/utils/pci.py

```diff
@@ -0,0 +1,310 @@
+
+    :return: specific pci id of a pci device.
+    """
+    cmd = "lspci -Dnvmm -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if part == line.split()[0].strip(':'):
+                return line.split()[-1]
+
+
+def get_pci_id(device):
+    """
+    Gets pci id of given divice.
+
+    :parm device: pci device.
+
+    :return: pci id of a pci device.
+    """
+    pci_id = []
+    for params in ['Vendor', 'Device', 'SVendor', 'SDevice']:
+        output = get_specific_pci_id(device, params)
+        if output:
+            pci_id.append(output)
+    if pci_id:
+        return ":".join(pci_id)
+
+
+def get_driver(device):
+    """
+    Gets the kernel driver in use of given device.
+
+    :parm device: pci device.
+
+    :return: driver of a pci device.
+    """
+    cmd = "lspci -ks %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if 'Kernel driver in use:' in line:
+                return line.split()[-1]
+
+
+def get_memory_address(device):
+    """
+    Gets the memory address of a pci device.
+    There may be multiple memory address for a pci device.
+    This function rerutns only the first such address.
+
+    :parm device: pci device.
+
+    :return: memory address of a pci device.
+    """
+    cmd = "lspci -bv -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if 'Memory at' in line:
+                return "0x%s" % line.split()[2]
+
+
+def get_mask(device):
+    """
+    Gets the mask of pci device.
+    There may be multiple memory entries for a pci device.
+    This mask is calculated only with the first such entry.
+
+    :parm device: pci device.
+
+    :return: mask of a pci device.
+    """
+    cmd = "lspci -vv -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        dic = {'K': '10', 'M': '20', 'G': '30'}
+        for line in output.splitlines():
+            if 'Region' in line and 'Memory at' in line:
+                val = line.split('=')[-1].split(']')[0]
+                memory_size = int(val[:-1]) * pow(2, int(dic[val[-1]]))
+                break
+        mask = hex((memory_size - 1) ^ int("0xffffffff", 16))
+        return mask
+
+
+def get_pci_devices():
+    """
+    Gets list of pci devices in the system.
+
+    :return: list of pci devices.
+    """
+    devices = []
+    cmd = "lspci -D"
+    for line in process.system_output(cmd, shell=True).splitlines():
+        if "PCI bridge" not in line:
+            devices.append(line.split()[0])
+    if devices:
+        return devices
+
+
+def get_vpd(device):
+    """
+    Gets the vpd of the given device.
+    lsvpd lists the Vital Product Data of a device.
```

Comment:
would you also please not about the availability? I found it in AIX documentation, so I guess it's available only there, or maybe on powerkvm?